internal/freelist/freelist.go

```diff
@@ -0,0 +1,109 @@
+package freelist
+
+import (
+	"sort"
+
+	"go.etcd.io/bbolt/internal/common"
+)
+
+type ReadWriter interface {
+	// Read calls Init with the page ids stored in te given page.
+	Read(page *common.Page)
+
+	// Write writes the freelist into the given page.
+	Write(page *common.Page)
+
+	// EstimatedWritePageSize returns the size of the freelist after serialization in Write.
+	// This should never underestimate the size.
+	EstimatedWritePageSize() int
+}
+
+type Interface interface {
+	ReadWriter
+
+	// Init initializes this freelist with the given list of pages.
+	Init(ids common.Pgids)
+
+	// Allocate returns the starting page id of a contiguous block of a given size in number of pages.
+	// If a contiguous block cannot be found then 0 is returned.
+	Allocate(txid common.Txid, numPages int) common.Pgid
+
+	// Count returns the number of free and pending pages.
+	Count() int
+
+	// FreeCount returns the number of free pages.
+	FreeCount() int
+
+	// FreePageIds returns all free pages.
+	FreePageIds() common.Pgids
+
+	// PendingCount returns the number of pending pages.
+	PendingCount() int
+
+	// AddReadonlyTXID adds a given read-only transaction id for pending page tracking.
+	AddReadonlyTXID(txid common.Txid)
+
+	// RemoveReadonlyTXID removes a given read-only transaction id from pending page tracking.
+	RemoveReadonlyTXID(txid common.Txid)
+
+	// FreePages releases any pages associated with closed read-only transactions.
+	FreePages()
+
+	// Free releases a page and its overflow for a given transaction id.
+	// If the page is already free then a panic will occur.
+	Free(txId common.Txid, p *common.Page)
+
+	// Freed returns whether a given page is in the free list.
+	Freed(pgId common.Pgid) bool
+
+	// Rollback removes the pages from a given pending tx.
+	Rollback(txId common.Txid)
+
+	// pendingPageIds returns all pending pages by transaction id.
+	pendingPageIds() map[common.Txid]*txPending
+
+	// release moves all page ids for a transaction id (or older) to the freelist.
+	release(txId common.Txid)
+
+	// releaseRange moves pending pages allocated within an extent [begin,end] to the free list.
+	releaseRange(begin, end common.Txid)
+}
+
+// Copyall copies a list of all free ids and all pending ids in one sorted list.
+// f.count returns the minimum length required for dst.
+func Copyall(f Interface, dst []common.Pgid) {
+	m := make(common.Pgids, 0, f.PendingCount())
+	for _, txp := range f.pendingPageIds() {
+		m = append(m, txp.ids...)
+	}
+	sort.Sort(m)
+	common.Mergepgids(dst, f.FreePageIds(), m)
+}
+
+// Reload reads the freelist from a page and filters out pending items.
+func Reload(f Interface, p *common.Page) {
+	f.Read(p)
+	NoSyncReload(f, p.FreelistPageIds())
+}
+
+// NoSyncReload reads the freelist from Pgids and filters out pending items.
+func NoSyncReload(f Interface, pgIds common.Pgids) {
```

Comment:
Can we turn these three functions into methods? Also Copyall is identical to `(t *shared) copyall`

Reply:
I found those easier to grasp with composition, but I can add those back into the shared interface as with copyall

Reply:
moved them back, replaced it with their old exact implementation in freelist.go